File: 061/rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
    if (!head || !head.next)
        return head;

    let node = head;
    let len = 1;
    while (node.next) {
        len += 1;
        node = node.next;
    }

    node.next = head
    k %= len

    for (let i = 0; i < len - k; i++)
        node = node.next;

    head = node.next;
    node.next = null;
    return head;
};

File: 061/RotateList.kt
class RotateList {

    fun rotateRight(head: ListNode?, k: Int): ListNode? {
        var anotherHead = head
        var anotherK = k
        if (anotherHead == null || anotherHead.next == null) {
            return head
        }

        var node = anotherHead
        var length = 1
        while (node?.next != null) {
            length++
            node = node.next
        }

        node?.next = anotherHead // Form a circle
        anotherK %= length

        for (i in 0 until length - anotherK) {
            node = node?.next
        }
        anotherHead = node?.next
        node?.next = null

        return anotherHead
    }

    data class ListNode(
            var `val`: Int,
            var next: ListNode? = null
    )

}